### PR TITLE
Feature/ub standalone dbengine

### PR DESCRIPTION
**Pull request by ubozov**
This commits includes the following changes:

1. standalone `postgresql` db engine is used
2. single `zip` archive for installation (includes `dappctrl`, `dappgui` and `dbengine`)

Resolves BV-859 and BV-867

### CODE REVIEW

File: dapp/dapp.go [hunk 1]
```diff
@@ -1,41 +1,188 @@
...
+	json.NewDecoder(read).Decode(&jsonMap)
+
+	if db, ok := jsonMap["DB"].(map[string]interface{}); ok {
+		if conn, ok := db["Conn"].(map[string]interface{}); ok {
+			conn["user"] = d.DBEngine.DB.User
+			if len(d.DBEngine.DB.Password) > 0 {
+				conn["password"] = d.DBEngine.DB.Password
+			}
+			conn["port"] = d.DBEngine.DB.Port
+			conn["dbname"] = d.DBEngine.DB.DBName
```
Comment (drew2a) [inline dapp/dapp.go]: These strings, I mean "password", "port" and "db name" has been used in different places. I suggest extract they to constants.

File: util/misc.go [hunk 2]
```diff
@@ -195,3 +197,130 @@ func ParseVersion(s string) int64 {
...
+			if err != nil {
+				return filenames, err
+			}
+			_, err = io.Copy(outFile, rc)
+			outFile.Close()
+
+			if err != nil {
+				return filenames, err
+			}
+		}
```
Comment (drew2a) [inline util/misc.go]: To me, personally, this is too big "else" block in too big "for" loop :)

Comment (ubozov) [reply util/misc.go]: updated

File: dapp/dapp.go (hunk 1 above)
Comment (ubozov) [reply dapp/dapp.go]: ok, I need to think about it

File: command/install.go [hunk 3]
```diff
@@ -25,196 +18,73 @@ func getInstallCmd() *installCmd {
...
-
-	// rename existing folder name to backup
-	if ok {
-		existDapp.Service.Uninstall()
-		newDapp.BackupPath = util.RenamePath(existDapp.InstallPath, "backup")
-		if err := os.Rename(existDapp.InstallPath, newDapp.BackupPath); err != nil {
-			return false, err
-		}
-		logger.Info("existing dapp version successfully backuped")
+		fmt.Println("for update you should run with command 'update'")
```
Comment (furkhat) [inline command/install.go]: `to update run 'update' command`

File: command/install.go [hunk 4]
```diff
@@ -25,196 +18,73 @@ func getInstallCmd() *installCmd {
...
+		logger.Warn("dapp was already installed")
```
Comment (furkhat) [inline command/install.go]: `dapp is installed`

File: command/update.go [hunk 5]
```diff
@@ -1,5 +1,146 @@
...
+	defer os.RemoveAll(conf.Dapp.TempPath)
+
+	return updateDapp(cmd, conf, logger)
+}
+
+func updateDapp(cmd *updateCmd, conf *config, logger log.Logger) error {
+	existDapp, ok := existingDapp(conf.Dapp.UserRole, logger)
+
+	if !ok {
+		fmt.Println("dapp is not exist.")
```
Comment (furkhat) [inline command/update.go]: Ambiguous. May be `dapp is not installed`?

File: command/update.go [hunk 6]
```diff
@@ -1,5 +1,146 @@
...
+		fmt.Println("for install you should run with command 'install'")
```
Comment (furkhat) [inline command/update.go]: `to install run 'install' command`

File: command/update.go [hunk 7]
```diff
@@ -1,5 +1,146 @@
...
+		logger.Warn("dapp is not exists")
```
Comment (furkhat) [inline command/update.go]: Ambiguous. May be `dapp is not installed`?

File: command/update.go [hunk 8]
```diff
@@ -1,5 +1,146 @@
...
+
+	newDapp, err := initDapp(conf)
+	if err != nil {
+		return err
+	}
+
+	version := util.ParseVersion(newDapp.Version)
+	if util.ParseVersion(existDapp.Version) >= version {
+		fmt.Println("you don't need to update. your dapp version:",
+			existDapp.Version)
```
Comment (furkhat) [inline command/update.go]: `fmt.Printf("dapp current version: %s, update is not required", existingDapp.Version)`

File: command/update.go [hunk 9]
```diff
@@ -1,5 +1,146 @@
...
+		logger.Warn("don't need to update")
```
Comment (furkhat) [inline command/update.go]: `already updated`

File: command/update.go [hunk 10]
```diff
@@ -1,5 +1,146 @@
...
+		return nil
+	}
+
+	configFile := filepath.Join(existDapp.InstallPath,
+		existDapp.Controller.Configuration)
+	if newDapp.DBEngine.DB, err = dbParamsFromConfig(configFile); err != nil {
+		logger.Warn(fmt.Sprintf("ocurred error when read config %v", err))
```
Comment (furkhat) [inline command/update.go]: `"failed to read config: %v"` or `"could not read config: %v"`, choose either of these styles and use it everywhere.

File: command/update.go [hunk 11]
```diff
@@ -1,5 +1,146 @@
...
+		return err
+	}
+
+	// update dapp core
+	if err := newDapp.Update(existDapp, logger); err != nil {
+
+		existDapp.DBEngine.Start(existDapp.InstallPath)
+		existDapp.Controller.Service.Start()
+
+		logger.Warn(fmt.Sprintf("ocurred error when update dapp %v", err))
```
Comment (furkhat) [inline command/update.go]: `"failed to read config: %v"` or `"could not read config: %v"`, choose either of these styles and use it everywhere.

**Comment by ubozov** [comment]
@furkhat , source codes were updated

File: command/update.go [hunk 12]
```diff
@@ -1,5 +1,146 @@
...
+
+	volume := filepath.VolumeName(conf.Dapp.InstallPath)
+	conf.Dapp.TempPath = util.TempPath(volume)
+	defer os.RemoveAll(conf.Dapp.TempPath)
+
+	return updateDapp(cmd, conf, logger)
+}
+
+func updateDapp(cmd *updateCmd, conf *config, logger log.Logger) error {
+	existDapp, ok := existingDapp(conf.Dapp.UserRole, logger)
```
Comment (furkhat) [inline command/update.go]: `oldDapp`.
`existDapp` sounds like a question.

File: dapp/dapp.go [hunk 13]
```diff
@@ -1,41 +1,188 @@
...
+	return &Dapp{
+		DBEngine: dbengine.NewConfig(),
+	}
 }
 
-// DownloadDappCtrl returns temporary downloaded path.
-func (d *Dapp) DownloadDappCtrl() string {
-	filePath := d.TempPath + path.Base(d.DownloadCtrl)
-	if err := util.DownloadFile(filePath, d.DownloadCtrl); err != nil {
+// DownloadDapp downloads dapp and returns temporary downloaded path.
```
Comment (furkhat) [inline dapp/dapp.go]: `DownloadDapp downloads dapp and returns temporary download path.`

File: dapp/dapp.go [hunk 14]
```diff
@@ -1,41 +1,188 @@
...
+	if err := util.DownloadFile(filePath, d.Download); err != nil {
 		return ""
 	}
 
 	return filePath
 }
+
+// Install installs a dapp core.
+func (d *Dapp) Install(logger log.Logger) error {
+	// install dbengine
```
Comment (furkhat) [inline dapp/dapp.go]: start comments with capital letter and end with period.

File: util/misc_unix.go [hunk 15]
```diff
@@ -15,14 +16,13 @@ func CheckSystemPrerequisites(volume string, logger log.Logger) bool {
...
-
 // ExistingDappCtrlVersion returns existing dappctrl version.
 func ExistingDappCtrlVersion(logger log.Logger) (string, bool) {
 	logger.Warn(fmt.Sprintf("your OS %s is not supported now", runtime.GOOS))
 	return "", false
 }
+
+// GrantAccess grants access to directory.
+func GrantAccess(path, user string) error {
+	return errors.New("does not implemented")
```
Comment (furkhat) [inline util/misc_unix.go]: just `not implemented`

File: command/command.go [hunk 16]
```diff
@@ -201,15 +142,42 @@ func existingDapp(role string, logger log.Logger) (*dapp.Dapp, bool) {
...
+	flag.CommandLine.Parse(os.Args[2:])
+
+	if *h {
+		help()
+		return true
+	}
+
+	if *configFile == "" {
+		logger.Warn("config parameter is empty")
+		fmt.Println("config parameter is empty")
```
Comment (furkhat) [inline command/command.go]: Our logger can log both to file and stdout. May be you could get rid of fmt prints and use only logger?

File: command/update.go (hunk 12 above)
Comment (ubozov) [reply command/update.go]: fixed

File: dapp/dapp.go (hunk 13 above)
Comment (ubozov) [reply dapp/dapp.go]: fixed

File: dapp/dapp.go (hunk 14 above)
Comment (ubozov) [reply dapp/dapp.go]: fixed

File: util/misc_unix.go (hunk 15 above)
Comment (ubozov) [reply util/misc_unix.go]: fixed

File: command/command.go (hunk 16 above)
Comment (ubozov) [reply command/command.go]: We use `fmt` to display more friendly messages.

**Pull request merged**